### PR TITLE
Change default epsilon_greedy to 0

### PR DESCRIPTION
**Pull request by hnyu**
It's easy to forget this hyperparameter when training a job. For almost all experiments, this should be 0.

### CODE REVIEW

**Comment by emailweixu** [comment]
I thought a nonzero epsilon_greedy is to prevent infinite evaluation due to deterministic behavior (e.g. breakout). Perhaps you need to explicitly set epsilon_greedy for those tasks.

**Comment by hnyu** [comment]
> I thought a nonzero epsilon_greedy is to prevent infinite evaluation due to deterministic behavior (e.g. breakout). Perhaps you need to explicitly set epsilon_greedy for those tasks.

So far, infinite loop only exists in Atari games. But it's easy to find out this kind of issue.

**Comment by emailweixu** [comment]
> > I thought a nonzero epsilon_greedy is to prevent infinite evaluation due to deterministic behavior (e.g. breakout). Perhaps you need to explicitly set epsilon_greedy for those tasks.
> 
> So far, infinite loop only exists in Atari games. But it's easy to find out this kind of issue.

Actually, I feel it will be hard to find out because there won't be any error message.

**Comment by hnyu** [comment]
> > > I thought a nonzero epsilon_greedy is to prevent infinite evaluation due to deterministic behavior (e.g. breakout). Perhaps you need to explicitly set epsilon_greedy for those tasks.
> > 
> > 
> > So far, infinite loop only exists in Atari games. But it's easy to find out this kind of issue.
> 
> Actually, I feel it will be hard to find out because there won't be any error message.

Both won't produce any error message. Using the default epsilon_greedy=0.1 but thinking it's 0 (at most papers set) is confusing. For an infinite loop, the episode length will be very large during evaluation and the training probably just stops proceeding; for playing, it's very easy to find out the loop.

**Comment by emailweixu** [comment]
> > > > I thought a nonzero epsilon_greedy is to prevent infinite evaluation due to deterministic behavior (e.g. breakout). Perhaps you need to explicitly set epsilon_greedy for those tasks.
> > > 
> > > 
> > > So far, infinite loop only exists in Atari games. But it's easy to find out this kind of issue.
> > 
> > 
> > Actually, I feel it will be hard to find out because there won't be any error message.
> 
> Both won't produce any error message. Using the default epsilon_greedy=0.1 but thinking it's 0 (at most papers set) is confusing. For an infinite loop, the episode length will be very large during evaluation and the training probably just stops proceeding; for playing, it's very easy to find out the loop.

So changing the gin file for atari to explicitly setting epsilon_greedy=0.0 seems a reasonable solution, as we know it is a problem for atari.

**Comment by hnyu** [comment]
> > > > > I thought a nonzero epsilon_greedy is to prevent infinite evaluation due to deterministic behavior (e.g. breakout). Perhaps you need to explicitly set epsilon_greedy for those tasks.
> > > > 
> > > > 
> > > > So far, infinite loop only exists in Atari games. But it's easy to find out this kind of issue.
> > > 
> > > 
> > > Actually, I feel it will be hard to find out because there won't be any error message.
> > 
> > 
> > Both won't produce any error message. Using the default epsilon_greedy=0.1 but thinking it's 0 (at most papers set) is confusing. For an infinite loop, the episode length will be very large during evaluation and the training probably just stops proceeding; for playing, it's very easy to find out the loop.
> 
> So changing the gin file for atari to explicitly setting epsilon_greedy=0.0 seems a reasonable solution, as we know it is a problem for atari.

updated

File: alf/trainers/policy_trainer.py [hunk 1]
```diff
@@ -718,8 +719,8 @@ def play(root_dir,
         logging.info(
             "%s: %s", m.name,
             map_structure(
-                lambda x: x.numpy().item() if x.ndim == 0 else x.numpy(),
-                m.result()))
+                lambda x: x.cpu().numpy().item()
+                if x.ndim == 0 else x.cpu().numpy(), m.result()))
```
Comment (emailweixu) [inline alf/trainers/policy_trainer.py:723]: It's strange that how the previous code can work.

**Review by emailweixu — APPROVED** [review]

File: alf/trainers/policy_trainer.py (hunk 1 above)
Comment (hnyu) [reply alf/trainers/policy_trainer.py:723]: > It's strange that how the previous code can work.

If play.py is called in CPU mode, then it won't be any problem. So I guess on CI it'll work but locally there will be problem.

**Pull request merged**